web3/datastructures.py

```diff
@@ -261,7 +261,7 @@ def __iter__(self) -> Iterator[TKey]:
 
     def __add__(self, other: Any) -> "NamedElementOnion[TKey, TValue]":
         if not isinstance(other, NamedElementOnion):
-            raise NotImplementedError(
+            raise NotImplemented(
```

Comment:
Thanks for the PR. This is the right idea but `NotImplemented` is not callable and is instead returned, rather than raised. Unfortunately we can't give a good message here. I vote for taking the message out and leaving it as a comment so that a curious user can at least see a reasoning if they navigate there.

```py
if not isinstance(other, NamedElementOnion):
    # you can only combine with another ``NamedElementOnion``
    return NotImplemented
```

Thoughts?

Reply:
@fselmo thank you for clearing out my confusion. I apologize for the mistake. I should have written `return` instead of `raise`. I think your suggestion would be the best.

Should I fix the issue and create another PR? Please let me know.